backend/regalloc/regalloc_split_state.ml

```diff
@@ -43,6 +43,176 @@ let log_renaming_info : indent:int -> t -> unit =
         Printmach.regset regset)
     state.phi_at_beginning
 
+(* Optimizes `destructions_at_end` and `definitions_at_beginning`, by moving
+   definitions down and destructions up. Doing so create more opportunities for
+   `RemoveReloadSpillInSameBlock`, and reduces the live ranges of temporaries,
+   hence making things slightly easier for the register allocator. *)
+module MoveSpillsAndReloads : sig
+  val optimize :
+    Cfg_with_infos.t ->
+    destructions_at_end:destructions_at_end ->
+    definitions_at_beginning:definitions_at_beginning ->
+    destructions_at_end * definitions_at_beginning
+end = struct
+  (** A definition at the begin of a block can be moved down if:
+
+     - the block does neither raise nor end with a destruction point;
+     - the block has only one (normal) successor;
+     - the successor has only one predecessor;
+     - the block and its successor are different;
+     - the defined register has no occurrences in the block. *)
+  let move_definitions_at_beginning_down :
+      Cfg_with_infos.t ->
+      definitions_at_beginning:definitions_at_beginning ->
+      definitions_at_beginning * Label.t Stack.t =
+   fun cfg_with_infos ~definitions_at_beginning ->
+    if split_debug
+    then log ~indent:1 "MoveSpillsAndReloads.move_definitions_at_beginning_down";
+    let definitions_at_beginning = ref definitions_at_beginning in
+    let doms = Cfg_with_infos.dominators cfg_with_infos in
+    let stack = Stack.create () in
+    Cfg_dominators.iter_breadth_dominator_tree
+      doms.Cfg_dominators.dominator_tree ~f:(fun (label : Label.t) ->
+        Stack.push label stack;
+        let block = Cfg_with_infos.get_block_exn cfg_with_infos label in
+        match Label.Map.find_opt label !definitions_at_beginning with
+        | None -> ()
+        | Some (definitions : Reg.Set.t) -> (
+          if Option.is_none (destruction_point_at_end block)
+             && Option.is_none block.exn
+          then
+            let successor_labels =
+              Cfg.successor_labels ~normal:true ~exn:false block
+            in
+            match Label.Set.cardinal successor_labels with
+            | 1 ->
+              let successor_label = Label.Set.choose successor_labels in
+              let successor_block =
+                Cfg.get_block_exn
+                  (Cfg_with_infos.cfg cfg_with_infos)
+                  successor_label
+              in
+              if (not (Label.equal label successor_label))
+                 && Label.Set.cardinal successor_block.predecessors = 1
+              then
+                let to_move : Reg.Set.t =
+                  Reg.Set.filter
+                    (fun (reg : Reg.t) ->
+                      let occurs = occurs_block block reg in
```

Comment:
Is it enough to check for reads, and a write means the definition can be removed?

Reply:
I don't think you can have a write without
seeing a read first, otherwise you would
not have a definition at a start of the block.

If you mean that as an optimization, we
could ignore `res`, I would rather leave a
CR-soon for now (the code of this module
is not optimal and I would rather do an
optimization pass after more testing).

Reply:
sounds good. yes, only meant as an optimization.